VisioShape.ps1

```diff
@@ -130,3 +130,202 @@ Function Get-VisioShape{
+    Param (
+	    [Parameter(Mandatory=$true)]$Master,
+	    [Parameter(Mandatory=$true)]$Label,
+	    [Parameter(Mandatory=$false)]$Name,
+	    [Parameter(Mandatory=$true)]$RackLabel,
+	    [Parameter(Mandatory=$true)]$RackVendor,
+	    [Parameter(Mandatory=$true)]$FirstU,
+		[Parameter(Mandatory=$false)]$X,
+		[Parameter(Mandatory=$false)]$Y
+    )
+    If($PSCmdlet.ShouldProcess('Visio','Drop a new rack equipment shape and connect to an existing rack shape')){
+        If($Master -is [string]){
+            $Master=$script:Shapes[$Master]
+        }
+        If(!$Name){
+            $Name=$Label
+        }
+
+        If(!$X){
+            $X=1.00
+        }
+
+        If(!$Y){
+            $Y=0.25
+        }
+
+        $p=Get-VisioPage
+        $ExistingRackShape=$p.Shapes | Where-Object {$_.Name -eq $RackLabel}
+        If(!$ExistingRackShape){
+            Write-Verbose "Existing rack shape $RackLabel was NOT found on the active page $($p.Name). Skipping!"
+            Break
+        }
+        Else {
+            Write-Verbose "Existing rack shape $RackLabel was found on the active page $($p.Name)."
+            If ($RackVendor -match 'HPE') {
+                $FirstUBXY = $FirstU | ForEach-Object { $_.ToString("00") }
+                $FirstUEXY = $FirstU | ForEach-Object { $_.ToString("00") }
+                $BeginXY = "=PAR(PNT($RackLabel!Connections.U$($FirstUBXY)B.X,$RackLabel!Connections.U$($FirstUBXY)B.Y))"
+                $EndXY = "=PAR(PNT($RackLabel!Connections.U$($FirstUEXY)E.X,$RackLabel!Connections.U$($FirstUEXY)E.Y))"
+            }
+            ElseIf ($RackVendor -match 'Dell') {
+                $FirstUBXY = $FirstU*2+3
+                $FirstUEXY = $FirstU*2+4
+                $BeginXY = "=PAR(PNT($RackLabel!Connections.X$($FirstUBXY),$RackLabel!Connections.Y$($FirstUBXY)))"
+                $EndXY = "=PAR(PNT($RackLabel!Connections.X$($FirstUEXY),$RackLabel!Connections.Y$($FirstUEXY)))"
+            }
+            ElseIf ($RackVendor -match 'IBM') {
+                $FirstUBXY = $FirstU*2+3
+                $FirstUEXY = $FirstU*2+4
+                $BeginXY = "=PAR(PNT($RackLabel!Connections.X$($FirstUBXY),$RackLabel!Connections.Y$($FirstUBXY)))"
+                $EndXY = "=PAR(PNT($RackLabel!Connections.X$($FirstUEXY),$RackLabel!Connections.Y$($FirstUEXY)))"
+            }
+            ElseIf ($RackVendor -match 'Cisco') {
+                $FirstUBXY = $FirstU
+                $FirstUEXY = $FirstU
+                $BeginXY = "=PAR(PNT($RackLabel!Connections.Cab$($FirstUBXY)A.X,$RackLabel!Connections.Cab$($FirstUBXY)A.Y))"
+                $EndXY = "=PAR(PNT($RackLabel!Connections.Cab$($FirstUBXY)B.X,$RackLabel!Connections.Cab$($FirstUBXY)B.Y))"
+            }
+            Else {
+                Write-Verbose "Rack Vendor $RackVendor is NOT known/supported. Skipping!"
+                Break
+            }
+
+        }
+
+<#
+    @Mike - unsure where $updatemode is set and if I need to worry about it here?

```

Comment:
So, UpdateMode is kind of unsupported and experimental (and cool).

I would probably use an if/then like this:
if($updateMode){
  $droppedshape=...  (as in the commented code)
} else {
  $droppedShape=$p.Drop .... (as you have it)
}

Reply:
Updated as per feedback in VisioShape.ps1 - https://github.com/mc1903/VisioBot3000/commit/b94cc8f28f60c43400718dcb713ed6a67ad6203d

Reply:
I should have been more clear.

In UpdateMode, to point is to use shapes on the page if they exist...if not, create a shape as necessary.

If you look at new-visioshape you can see how this might look.

As is, you wouldn't update the position of a rack shape in update mode.

Probably not a huge deal, but if we can make it work it would be better.

Reply:
> _I should have been more clear..._

No, I am probably being a bit thick :-). 

I assume I just set `$UpdateMode=$true` before calling `New-VisioRackShape` with updated rack equipment shape BeginXY and EndXY placements?

Again, I will look to fix this later in the week/weekend. Thx.